Improve error reporting: Missing = on type declaration

tboby opened the pull request:
This is my attempt to implement #1121, leaning on the start made in #6603.

This is my first contribution, so feedback appreciated!

I'm not quite sure what's happening with the xlf files. Is it possible a previous commit on master forgot to generate them?

smoothdeveloper commented:
@tboby
> I'm not quite sure what's happening with the xlf files. Is it possible a previous commit on master forgot to generate them?

It is better to consider they are in .gitignore (despite they aren't), and not check them in unless explicitly asked during the review.

abelbraaksma commented:
Not 100% sure, but maybe the xlf files had something to do with: https://github.com/dotnet/fsharp/pull/9641#pullrequestreview-444181842

I do know that they need to be rebuild when texts change, so that the translation team gets notified to translate the new texts. That that means that there's a bunch of changes at once seems to me to be expected.

TIHan commented on src/fsharp/pars.fsy:
```diff
@@ -5392,6 +5393,10 @@ deprecated_opt_equals:
   | EQUALS      { deprecatedWithError (FSComp.SR.parsNoEqualShouldFollowNamespace()) (lhs parseState); () } 
   | /* EMPTY */ {  }
 
+type_opt_equals:
```
I would just name this `opt_equals`.

TIHan commented on src/fsharp/FSComp.txt:
```diff
@@ -1492,6 +1492,7 @@ notAFunctionButMaybeDeclaration,"This value is not a function and cannot be appl
 3350,chkFeatureNotLanguageSupported,"Feature '%s' is not available in F# %s. Please use language version %s or greater."
 3351,chkFeatureNotRuntimeSupported,"Feature '%s' is not supported by target runtime."
 3352,typrelInterfaceMemberNoMostSpecificImplementation,"Interface member '%s' does not have a most specific implementation."
+3360,parsEqualsMissingInTypeDefinition,"Unexpected symbol in type definition. Did you forget to use the = operator?"
```
*Unexpected token

TIHan replied:
Acually, "Unexpected token in type definition. Expected '='."

tboby commented:
I've:
- rebased onto main
- removed the xlf files as I wasn't sure if I was supposed to leave them in
- updated the error message and parser naming as suggested

cartermp commented on src/fsharp/FSComp.txt:
```diff
@@ -1498,6 +1498,7 @@ notAFunctionButMaybeDeclaration,"This value is not a function and cannot be appl
 3351,chkFeatureNotRuntimeSupported,"Feature '%s' is not supported by target runtime."
 3352,typrelInterfaceMemberNoMostSpecificImplementation,"Interface member '%s' does not have a most specific implementation."
 3353,chkFeatureNotSupportedInLibrary,"Feature '%s' requires the F# library for language version %s or greater."
+3360,parsEqualsMissingInTypeDefinition,"Unexpected token in type definition. Expected '='."
```
I think "Unexpected token in type definition. Expected '=' after the type name" would be great, and even better (if it's possible to pass the type name) would be "after the type 'type-name'", where `type-name` is passed in from the parser.

But if it's not possible to pass in the name then that's fine. Otherwise this is looking good!

tboby replied:
I've added the type name as well, might as well!
I think I've found an error in the parser implementation of the spec to do with this ident though, which I'll write up below.

tboby commented:
I've amended this to include the type name in the error message.
However, this required me to take the head or tail of the list of idents, which was odd, as the spec implies a type def's type name should be an ident:
`type-name : attributesopt accessopt ident typar-defnsopt` 
not a `long-ident` (which includes dots).

The `tyconDefn:` parser definition I edit here eventually just uses `path` which is `SyntaxTree.LongIdentWithDots`. If you try and use dots in a type name you do seem to get an error, but this is caught by the TypeChecker which emits "Invalid type extension".

I may be missing something, but unless someone lets me know better I will write this up into an issue and consider having a go at changing this. I assume it is better to catch a spec error in the parser than in the type checker?

cartermp commented:
> I assume it is better to catch a spec syntax error in the parser than in the type checker?

Yes, absolutely :)

@dsyme any thoughts on the spec-y question?

cartermp reviewed (approved):
Thanks for this work, it is great!
